wscl-issues/draft/sequence-definition

```diff
@@ -0,0 +1,69 @@
+Issue:          SEQUENCE-DEFINITION
+Forum:          Cleanup
+Category:       CLARIFICATION
+Status:         draft
+Edit History:   16-May-24, Version 1 by Jin-Cheng Guu
+References:     SEQUENCE
+
+Problem Description:
+
+  In the draft ANSI Common Lisp specification, the description of the system
+  class SEQUENCE [a] conflicts with the description of the Sequence Concepts
+  [b], where
+
+    [a]: The types vector and the type list are disjoint subtypes of type
+    sequence, but are not necessarily an exhaustive partition of sequence.
+
+    [b]: A sequence is an ordered collection of elements, implemented as
+    either a vector or a list.
+
+Proposal:
+
+  This proposal changes the description of the Sequence Concepts so that [b]
+  instead reads:
+
+    > A sequence is an ordered collection of elements, implemented either as a
+    vector, a list, or possibly something else.
+
+Test Cases:
+
+  N/A (Test cases are difficult to provide, because conformating
```

Comment:
`(subtypep 'sequence '(or vector list))` could return `T NIL` or `NIL NIL` on an implementation conforming to this proposal, but `NIL T` would indicate nonconformance.

Also: "conforming"

Reply:
I'm not sure about this.. For a conforming implementation, we do need `(subtypep '(or vector list) 'sequence)` to return `T T`. However, the point of this issue is to give such an implementation the freedom to have something else in the type `sequence`.

Reply:
You read the arguments backwards. I'm saying you can test whether `sequence` is a subtype of `(or vector list)`, not the other way around. If any non-list non-vector sequences are possible, `sequence` is not a subtype of `(or vector list)`.

Of course I suppose this proposal doesn't necessarily **require** implementations to have non-vector non-list sequences, so maybe the subtype is inappropriate after all. Considered that way, having test cases at all would be inappropriate, since it's just making the standard more flexible without requiring any implementation to make any changes whatsoever.

Reply:
I am still confused. Why would `(subtypep 'sequence '(or vector list))`'s returning `NIL T` indicate nonconformance to this proposal? This proposal aims to provide the freedom for an implementation to have `sequence` to **possibly** have things more than just vectors and lists.

Reply:
A value of `NIL, T` or `NIL, NIL`  indicates the implementation does not conform to the current [b].

For this issue, the test cases would not indicate conformance or non-conformance, but indicate whether LIST and VECTOR are an exhaustive for the typespace of SEQUENCE. Supporting extensible sequences would require returning `NIL, T`, I would think.

```
Test Cases:

  (defun one ()
    (subtypep 'sequence '(or vector list)))

  (one) ; => T, T

Rationale:

  This is a plain contradiction in dpANS.

Current Practice:

  ABCL 1.9.3-dev-fasl43
    (one) ; => NIL, NIL

  SBCL 2.4.1
    (one) ; => NIL, T

  CLASP cclasp-boehmprecise-2.5.0-247-g4b148619f-g85ffab47-cst
    (one) ; => T, T

  ECL 23.9.9-f477a3ef
    (one) ; => T, T

  CMU 2024-03-10 18:28:32 (21E Unicode)
    (one) ; => T, T

  CCL 1.12-f98
    (one) ; => T, T
```